core/src/main/scala/com/ccadllc/cedi/dtrace/TraceT.scala

```diff
@@ -295,39 +397,51 @@ object TraceT extends TraceTPolyFunctions with TraceTInstances {
-   *       callback for signaling the result once it is ready
-   */
-  def async[F[_], A](cb: (Either[Throwable, A] => Unit) => Unit)(implicit F: Async[F]): TraceT[F, A] = toTraceT(F.async(cb))
-
-  /**
-   * Creates a cancelable `TraceT[F, A]` instance that executes an
-   * asynchronous process on evaluation.
+   * Asynchronous boundary described as an effectful `TraceT[F, Unit managed
+   * by the provided `ContextShift`.
    *
-   * This builder accepts a registration function that is
-   * being injected with a side-effectful callback, to be called
-   * when the asynchronous process is complete with a final result.
+   * This operation can be used in `flatMap` chains to "shift" the
+   * continuation of the run-loop to another thread or call stack.
    */
-  def cancelable[F[_], A](k: (Either[Throwable, A] => Unit) => IO[Unit])(implicit F: Concurrent[F]): TraceT[F, A] = toTraceT(F.cancelable(k))
+  def shift[F[_]](implicit cs: ContextShift[TraceT[F, ?]]): TraceT[F, Unit] = cs.shift
 
   /**
-   * Defines a conversion from [[IO]] in terms of the `Concurrent` type class.
+   * Lifts the non-strict, possibly impure expression computing a `TraceT[F, A]` into a `TraceT[F, A]`.
+   * The expression is suspended until the outer `TraceT` returned is run.
+   * @param t the non-strict expression computing `TraceT[F, A]` to lift into a `TraceT` context suspended
+   *   until the outer `TraceT` is run.
+   * @return a non-strict expression which computes `TraceT[F, A]` lifted into a `TraceT` in
+   *   a suspended state until the outer `TraceT` is run.
    */
-  def liftIO[F[_], A](ioa: IO[A])(implicit F: Concurrent[F]): TraceT[F, A] = toTraceT(F.liftIO(ioa))
+  def suspend[F[_], A](t: => TraceT[F, A])(implicit F: Sync[F]): TraceT[F, A] =
+    toTraceT(F.delay(t)).flatMap(identity)
 
   /**
    * Lifts a program `F` which computes `A` into a `TraceT[F, A]` context.
    * @param fa a program `F` which computes a value `A`.
    * @return a `TraceT[F, A]`
    */
   def toTraceT[F[_], A](fa: F[A]): TraceT[F, A] = TraceT { _ => fa }
+
+  /**
+   * Internal API — suspends the execution of `toEffect` in the `F` context.
+   *
+   * Used to build `TraceT[F, ?]` values for `F[_]` data types that implement `Monad`,
+   * in which case it is safer to trigger the `F[_]` context earlier.
+   *
+   * The basic requirement of `F` for callers of this function is `Functor`; however we are
+   * doing discrimination based on inheritance and if we detect a
+   * `Monad`, then we use it to trigger the `F[_]` context earlier.
+   *
+   * Triggering the `F[_]` context earlier is important to avoid stack
+   * safety issues for `F` monads that have stack safe implementations.
+   * For example `Eval` or `IO`. Without this the `Monad`
+   * instance is stack unsafe, even if the underlying `F` is stack safe.
+   */
+  private[dtrace] def suspendEffect[F[_], A](toEffect: TraceContext[F] => F[A])(implicit F: Functor[F]): TraceT[F, A] = F match {
```

Comment:
Why can't we take ` Monad[F]` constraint instead? Or a `StackSafeMonad[F]` even, if we rely on `flatMap` stack safety?

Reply:
Cool -- can you add a link to Alex's version?

Reply:
oh, shoot - no, you're right.  For the map case, I can just wrap in TraceT - since that's all suspendEffect is doing in that case.  Not sure why it was special cased in Kleisli.

Reply:
no, get a stackoverflow in tests when i change it - map needs suspend effect to be stacksafe (so in essence it needs to have a Monad, though we don't want to require that constraint at compile-time for map). Because map applies the TraceContext[F] => F[?] so that it can map over the value, so it's in essence a curried flatmap in disguise.  I'll add a FIXME to look at further and at least add a note to the scaladoc for map.

Reply:
I could just require Monad for map, but that breaks down when implementing the typeclasses ... sigh.

Reply:
No worries. Let's keep it with a comment linking to cats-effect.